producer_test.go

```diff
@@ -394,6 +394,91 @@ func TestProducerBrokerBounce(t *testing.T) {
 	safeClose(t, client)
 }
 
+func TestProducerMultipleRetries(t *testing.T) {
+	broker1 := NewMockBroker(t, 1)
+	broker2 := NewMockBroker(t, 2)
+	broker3 := NewMockBroker(t, 3)
+
+	response1 := new(MetadataResponse)
+	response1.AddBroker(broker2.Addr(), broker2.BrokerID())
+	response1.AddTopicPartition("my_topic", 0, broker2.BrokerID(), nil, nil, NoError)
+	broker1.Returns(response1)
+
+	client, err := NewClient("client_id", []string{broker1.Addr()}, nil)
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	config := NewProducerConfig()
+	config.FlushMsgCount = 10
+	config.AckSuccesses = true
+	config.MaxRetries = 4
+	config.RetryBackoff = 0
+	producer, err := NewProducer(client, config)
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	for i := 0; i < 10; i++ {
+		producer.Input() <- &MessageToSend{Topic: "my_topic", Key: nil, Value: StringEncoder(TestMessage)}
+	}
+	response2 := new(ProduceResponse)
+	response2.AddTopicPartition("my_topic", 0, NotLeaderForPartition)
+	broker2.Returns(response2)
+
+	response3 := new(MetadataResponse)
+	response3.AddBroker(broker3.Addr(), broker3.BrokerID())
+	response3.AddTopicPartition("my_topic", 0, broker3.BrokerID(), nil, nil, NoError)
+	broker1.Returns(response3)
+	broker3.Returns(response2)
+	broker1.Returns(response1)
+	broker2.Returns(response2)
+	broker1.Returns(response1)
+	broker2.Returns(response2)
+	broker1.Returns(response3)
```

Comment:
So, this tests the following sequence?

- broker 1 (metadata broker) tells us the partition lives on broker 2.
- broker 2 (produce attempt 1) tells us that is not actually the case
- broker 1 (metadata broker) tells us the partition now lives on broker 3.
- broker 3 (produce attempt 2) tells us this is not the case either.
- broker 1 (metadata broker) tells us the partition now lives on broker 2 again.
- broker 2 (produce attempt 3) tells us this is not the case
- broker 1 (metadata broker) tells us the partition now lives on broker 2 again.
- broker 2 (produce attempt 4) tells us this is still not the case
- broker 1 (metadata broker) tells us the partition now lives on broker 3.
- broker 4 (produce attempt 5, retry 4) produces successfully.

So this flow exactly fits in 4 retries.